File: ruby/binary_search/bsearch3.rb
require "test/unit"

module BSearch3
  def search value, values
    from = 0
    to = values.size

    while from < to
      mid_index = (from + to) / 2
      mid_value = values[mid_index]

      return mid_index if value == mid_value
      to = mid_index if value < mid_value
      from = mid_index + 1 if value > mid_value
    end

    return -1
  end
end

class TestBSearch3 < Test::Unit::TestCase
  include BSearch3

  def test_binary_search
    assert_equal -1, search(0, [])

    assert_equal -1, search(1, [2])
    assert_equal 0, search(2, [2])
    assert_equal -1, search(3, [2])

    assert_equal -1, search(1, [2, 3])
    assert_equal 0, search(2, [2, 3])
    assert_equal 1, search(3, [2, 3])
    assert_equal -1, search(4, [2, 3])
  end
end

File: kotlin/src/katas/kotlin/binarysearch/BSearch3.kt
package katas.kotlin.binarysearch

import kotlincommon.test.shouldEqual
import org.junit.Test

class BSearch3 {
    @Test fun `find index of an element in list`() {
        binarySearch(0, emptyList()) shouldEqual -1

        binarySearch(0, listOf(1)) shouldEqual -1
        binarySearch(1, listOf(1)) shouldEqual 0
        binarySearch(2, listOf(1)) shouldEqual -1

        binarySearch(0, listOf(1, 2)) shouldEqual -1
        binarySearch(1, listOf(1, 2)) shouldEqual 0
        binarySearch(2, listOf(1, 2)) shouldEqual 1
        binarySearch(3, listOf(1, 2)) shouldEqual -1

        binarySearch(0, listOf(1, 2, 3)) shouldEqual -1
        binarySearch(1, listOf(1, 2, 3)) shouldEqual 0
        binarySearch(2, listOf(1, 2, 3)) shouldEqual 1
        binarySearch(3, listOf(1, 2, 3)) shouldEqual 2
        binarySearch(4, listOf(1, 2, 3)) shouldEqual -1

        binarySearch(0, listOf(1, 2, 3, 4)) shouldEqual -1
        binarySearch(1, listOf(1, 2, 3, 4)) shouldEqual 0
        binarySearch(2, listOf(1, 2, 3, 4)) shouldEqual 1
        binarySearch(3, listOf(1, 2, 3, 4)) shouldEqual 2
        binarySearch(4, listOf(1, 2, 3, 4)) shouldEqual 3
        binarySearch(5, listOf(1, 2, 3, 4)) shouldEqual -1
    }
    
    private fun <T : Comparable<T>> binarySearch(value: T, list: List<T>): Int {
        var fromIndex = 0
        var toIndex = list.size
        while (fromIndex < toIndex) {
            val midIndex = (toIndex + fromIndex) / 2
            val midValue = list[midIndex]
            if (value == midValue) return midIndex
            else if (value < midValue) toIndex = midIndex
            else if (value > midValue) fromIndex = midIndex + 1
        }
        return -1
    }
}